deps/ncrypto/ncrypto.cc

```diff
@@ -2044,4 +2197,417 @@ Result<BIOPointer, bool> EVPKeyPointer::writePublicKey(
+                                        "tls_aes_128_ccm_sha256"};
+
+  const int n = sk_SSL_CIPHER_num(ciphers);
+
+  for (int i = 0; i < n; ++i) {
+    const SSL_CIPHER* cipher = sk_SSL_CIPHER_value(ciphers, i);
+    cb(SSL_CIPHER_get_name(cipher));
+  }
+
+  for (unsigned i = 0; i < 5; ++i) {
+    const char* name = TLS13_CIPHERS[i];
+    cb(name);
+  }
+}
+
+bool SSLPointer::setSession(const SSLSessionPointer& session) {
+  if (!session || !ssl_) return false;
+  return SSL_set_session(get(), session.get()) == 1;
+}
+
+bool SSLPointer::setSniContext(const SSLCtxPointer& ctx) const {
+  if (!ctx) return false;
+  auto x509 = ncrypto::X509View::From(ctx);
+  if (!x509) return false;
+  EVP_PKEY* pkey = SSL_CTX_get0_privatekey(ctx.get());
+  STACK_OF(X509) * chain;
+
+  int err = SSL_CTX_get0_chain_certs(ctx.get(), &chain);
+  if (err == 1) err = SSL_use_certificate(get(), x509);
+  if (err == 1) err = SSL_use_PrivateKey(get(), pkey);
+  if (err == 1 && chain != nullptr) err = SSL_set1_chain(get(), chain);
+  return err == 1;
+}
+
+std::optional<uint32_t> SSLPointer::verifyPeerCertificate() const {
+  if (!ssl_) return std::nullopt;
+  if (X509Pointer::PeerFrom(*this)) {
+    return SSL_get_verify_result(get());
+  }
+
+  const SSL_CIPHER* curr_cipher = SSL_get_current_cipher(get());
+  const SSL_SESSION* sess = SSL_get_session(get());
+  // Allow no-cert for PSK authentication in TLS1.2 and lower.
+  // In TLS1.3 check that session was reused because TLS1.3 PSK
+  // looks like session resumption.
+  if (SSL_CIPHER_get_auth_nid(curr_cipher) == NID_auth_psk ||
+      (SSL_SESSION_get_protocol_version(sess) == TLS1_3_VERSION &&
+       SSL_session_reused(get()))) {
+    return X509_V_OK;
+  }
+
+  return std::nullopt;
+}
+
+const std::string_view SSLPointer::getClientHelloAlpn() const {
+  if (ssl_ == nullptr) return {};
+  const unsigned char* buf;
+  size_t len;
+  size_t rem;
+
+  if (!SSL_client_hello_get0_ext(
+          get(),
+          TLSEXT_TYPE_application_layer_protocol_negotiation,
+          &buf,
+          &rem) ||
+      rem < 2) {
+    return {};
+  }
+
+  len = (buf[0] << 8) | buf[1];
+  if (len + 2 != rem) return {};
+  return reinterpret_cast<const char*>(buf + 3);
+}
+
+const std::string_view SSLPointer::getClientHelloServerName() const {
+  if (ssl_ == nullptr) return {};
+  const unsigned char* buf;
+  size_t len;
+  size_t rem;
+
+  if (!SSL_client_hello_get0_ext(get(), TLSEXT_TYPE_server_name, &buf, &rem) ||
+      rem <= 2) {
+    return {};
+  }
+
+  len = (*buf << 8) | *(buf + 1);
+  if (len + 2 != rem) return {};
+  rem = len;
+
+  if (rem == 0 || *(buf + 2) != TLSEXT_NAMETYPE_host_name) return {};
+  rem--;
+  if (rem <= 2) return {};
+  len = (*(buf + 3) << 8) | *(buf + 4);
+  if (len + 2 > rem) return {};
+  return reinterpret_cast<const char*>(buf + 5);
+}
+
+std::optional<const std::string_view> SSLPointer::GetServerName(
+    const SSL* ssl) {
+  if (ssl == nullptr) return std::nullopt;
```

Comment:
Nit: !ssl also works, and is more readable (open to debate ofc)

Reply:
I generally prefer explicit `== nullptr` checks for raw pointers.